### PR TITLE
Added Lambda EC2 Autoscaler Function configuration and updated README

### PR DESCRIPTION
**Pull request by eminalemdar**
## Description of the change

I've added the Lambda EC2 Autoscaler function to automatically make scale out or scale in decisions.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue);
- [x] New feature (non-breaking change that adds functionality);
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected);
- [ ] Documentation (a documentation or example fix not affecting the infrastructure managed by this module);

### Development

- [x] All necessary variables have been defined, with defaults if applicable;
- [x] The code is formatted properly;

### Code review

- [ ] The module version is bumped accordingly;
- [ ] Spacelift tests are passing;
- [x] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached;
- [ ] This pull request is no longer marked as "draft";
- [x] Reviewers have been assigned;
- [ ] Changes have been reviewed by at least one other engineer;

### CODE REVIEW

**Comment by jmfontaine** [comment]
@eminalemdar Please make sure that all tests pass before asking for a review. Thanks.

File: lambda.tf [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  runtime                        = "go1.x"
+
+  environment {
+    variables = {
+      AUTOSCALING_GROUP_ARN         = module.asg.autoscaling_group_arn
+      AUTOSCALING_REGION            = data.aws_region.this.name
+      SPACELIFT_API_KEY_ID          = var.spacelift_api_key_id
+      SPACELIFT_API_KEY_SECRET_NAME = aws_ssm_parameter.spacelift_api_key_secret.name
+      SPACELIFT_API_KEY_ENDPOINT    = var.spacelift_url
```
Comment (jmfontaine) [inline lambda.tf:34]:
```suggestion
      SPACELIFT_API_KEY_ENDPOINT    = var.spacelift_api_key_endpoint
```

See comment below.

File: lambda.tf [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  depends_on  = [ null_resource.download ]
+}
+
+resource "aws_lambda_function" "autoscaler" {
+  filename         = data.archive_file.binary.output_path
+  source_code_hash = data.archive_file.binary.output_base64sha256
+  function_name    = local.function_name
+  role             = aws_iam_role.lambda.arn
+  handler          = "ec2-workerpool-autoscaler_v${var.autoscaler_version}"
+
```
Comment (jmfontaine) [inline lambda.tf:25]:
```suggestion
```

File: lambda.tf [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  runtime                        = "go1.x"
```
Comment (jmfontaine) [inline lambda.tf:26]:
```suggestion
  runtime          = "go1.x"
```

File: lambda.tf [hunk 4]
```diff
@@ -0,0 +1,63 @@
+locals {
+  code_version  = var.autoscaler_version
```
Comment (jmfontaine) [inline lambda.tf:2]: It does not seem necessary since it is used only once and uses the variable value as-is.

File: lambda.tf [hunk 5]
```diff
@@ -0,0 +1,63 @@
+locals {
```
Comment (jmfontaine) [inline lambda.tf:1]: I suggest renaming this file into `autoscaler` and having everything related in there.

File: api_secret.tf [hunk 6]
```diff
@@ -0,0 +1,5 @@
+resource "aws_ssm_parameter" "spacelift_api_key_secret" {
```
Comment (jmfontaine) [inline api_secret.tf:1]: Does it deserve its own file? Feels more like a part of the autoscaler to me.

File: iam.tf [hunk 7]
```diff
@@ -33,3 +33,86 @@ resource "aws_iam_instance_profile" "this" {
   name = local.namespace
   role = var.create_iam_role ? aws_iam_role.this[0].name : var.iam_role_arn
 }
+
+data "aws_iam_policy_document" "lambda_policy" {
```
Comment (jmfontaine) [inline iam.tf:37]:
```suggestion
data "aws_iam_policy_document" "autoscaler" {
```

File: iam.tf [hunk 8]
```diff
@@ -33,3 +33,86 @@ resource "aws_iam_instance_profile" "this" {
...
+      "autoscaling:SetDesiredCapacity",
+    ]
+
+    resources = [module.asg.autoscaling_group_arn]
+  }
+
+  statement {
+    effect    = "Allow"
+    actions   = ["autoscaling:DescribeAutoScalingGroups"]
+    resources = ["*"]
```
Comment (jmfontaine) [inline iam.tf:75]: Can we limit it to `module.asg.autoscaling_group_arn` as above?

File: iam.tf [hunk 9]
```diff
@@ -33,3 +33,86 @@ resource "aws_iam_instance_profile" "this" {
...
+    effect = "Allow"
+    actions = [
+      "autoscaling:DetachInstances",
+      "autoscaling:SetDesiredCapacity",
+    ]
+
+    resources = [module.asg.autoscaling_group_arn]
+  }
+
+  statement {
```
Comment (jmfontaine) [inline iam.tf:72]: For the sake of consistency, this statement should be commented.

File: iam.tf [hunk 10]
```diff
@@ -33,3 +33,86 @@ resource "aws_iam_instance_profile" "this" {
...
+
+  # Allow the Lambda to read the secret from SSM Parameter Store.
+  statement {
+    effect    = "Allow"
+    actions   = ["ssm:GetParameter"]
+    resources = [aws_ssm_parameter.spacelift_api_key_secret.arn]
+  }
+}
+
+resource "aws_iam_role" "lambda" {
```
Comment (jmfontaine) [inline iam.tf:97]:
```suggestion
resource "aws_iam_role" "autoscaler" {
```

File: lambda.tf [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  function_name    = local.function_name
+  role             = aws_iam_role.lambda.arn
+  handler          = "ec2-workerpool-autoscaler_v${var.autoscaler_version}"
+
+  runtime                        = "go1.x"
+
+  environment {
+    variables = {
+      AUTOSCALING_GROUP_ARN         = module.asg.autoscaling_group_arn
+      AUTOSCALING_REGION            = data.aws_region.this.name
```
Comment (jmfontaine) [inline lambda.tf:31]: Can't the autoscaler determine the region it runs in automatically?

File: lambda.tf [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    mode = "Active"
+  }
+
+  depends_on = [ module.asg, null_resource.download ]
+}
+
+resource "aws_cloudwatch_event_rule" "scheduling" {
+  name                = "spacelift-${var.worker_pool_id}-scheduling"
+  description         = "Spacelift autoscaler scheduling for worker pool ${var.worker_pool_id}"
+  schedule_expression = "rate(1 minute)"
```
Comment (jmfontaine) [inline lambda.tf:49]: That should probably be configurable.

**Review by jmfontaine — CHANGES_REQUESTED** [review]

File: lambda.tf (hunk 12 above)
Comment (eminalemdar) [reply lambda.tf:49]: this event rule doesn't work with variable parameters unfortunately

File: README.md [hunk 13]
```diff
@@ -2,6 +2,8 @@
 
 Terraform module deploying a Spacelift worker pool on AWS EC2 using an autoscaling group.
 
+This module also deploys a Lambda function to check Spacelift and make scale in or scale out decisions. The function will check the queues in according Worker Pool and based on that increase or decrease the number of the workers.
```
[the remark this answers is not in the document]
Comment (eminalemdar) [reply README.md]: Done

File: variables.tf [hunk 14]
```diff
@@ -131,3 +131,30 @@ variable "instance_refresh" {
...
+  description = "ID of the Spacelift API key to use"
+}
+
+variable "spacelift_api_key_secret" {
+  type        = string
+  sensitive   = true
+  description = "Secret corresponding to the Spacelift API key to use"
+}
+
+variable "spacelift_url" {
```
[the remark this answers is not in the document]
Comment (eminalemdar) [reply variables.tf]: Done

File: logging.tf [hunk 15]
```diff
@@ -0,0 +1,4 @@
+resource "aws_cloudwatch_log_group" "log_group" {
```
[the remark this answers is not in the document]
Comment (eminalemdar) [reply logging.tf]: Done

File: lambda.tf [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+resource "aws_cloudwatch_event_target" "scheduling" {
+  rule = aws_cloudwatch_event_rule.scheduling.name
+  arn  = aws_lambda_function.autoscaler.arn
+}
+
+resource "aws_lambda_permission" "allow_cloudwatch_to_call_check_foo" {
```
[the remark this answers is not in the document]
Comment (eminalemdar) [reply lambda.tf]: Changed the resource name

File: lambda.tf (hunk 1 above)
Comment (eminalemdar) [reply lambda.tf]: Done

File: lambda.tf (hunk 2 above)
Comment (eminalemdar) [reply lambda.tf]: Done

File: lambda.tf (hunk 3 above)
Comment (eminalemdar) [reply lambda.tf]: Done

File: lambda.tf (hunk 4 above)
Comment (eminalemdar) [reply lambda.tf]: Done

File: lambda.tf (hunk 5 above)
Comment (eminalemdar) [reply lambda.tf]: Done

File: iam.tf (hunk 7 above)
Comment (eminalemdar) [reply iam.tf]: Done

File: iam.tf (hunk 8 above)
Comment (eminalemdar) [reply iam.tf]: Done

File: iam.tf (hunk 10 above)
Comment (eminalemdar) [reply iam.tf]: Done

**Comment by eminalemdar** [comment]
> @eminalemdar Please make sure that all tests pass before asking for a review. Thanks.

@jmfontaine the tests now pass. can you please review these changes again?

File: variables.tf [hunk 17]
```diff
@@ -131,3 +131,30 @@ variable "instance_refresh" {
 locals {
   namespace = "sp5ft-${var.worker_pool_id}"
 }
+
+variable "autoscaler_version" {
+  type        = string
+  description = "Version of the autoscaler to deploy"
+  default     = "0.1.0"
```
[the remark this answers is not in the document]
Comment (eminalemdar) [reply variables.tf:138]: @jmfontaine I've tried using `latest` but I think the version in the binary release is not latest that's why I've reverted my changes to the hardcoded version for now.

[the remark this answers is not in the document]
Comment (jmfontaine) [reply variables.tf:138]: No, `latest` does not exist but it should. You can submit a PR against the other repository to make it happen.

File: lambda.tf (hunk 11 above)
Comment (jmfontaine) [reply lambda.tf]: That was my point. There is no good reason for that, but let's avoid updating the autoscaler for now.

File: lambda.tf (hunk 12 above)
Comment (jmfontaine) [reply lambda.tf]: Are you saying that the following would not work?

```
schedule_expression = var.schedule_expression
```

The `schedule_expression` variable would default to `rate(1 minute)` but that would also allow people to tweak that.

File: variables.tf [hunk 18]
```diff
@@ -131,3 +131,34 @@ variable "instance_refresh" {
...
+}
+
+variable "spacelift_api_key_id" {
+  type        = string
+  description = "ID of the Spacelift API key to use"
+  default     = "0VOYU49U485BMZZVAWXU59VOW2"
```
Comment (jmfontaine) [inline variables.tf:144]: That should not be a default value for the variable but should be passed by the examples to the module.

File: variables.tf [hunk 19]
```diff
@@ -131,3 +131,34 @@ variable "instance_refresh" {
...
+}
+
+variable "spacelift_api_key_secret" {
+  type        = string
+  sensitive   = true
+  description = "Secret corresponding to the Spacelift API key to use"
+  default     = "f7anuofh4b6a4e43aplqt49099606de2mzbq4391tj1d3dc9872q23z8fvctu4kh"
```
Comment (jmfontaine) [inline variables.tf:151]: That should not be a default value for the variable but should be passed by the examples to the module.

File: variables.tf [hunk 20]
```diff
@@ -131,3 +131,34 @@ variable "instance_refresh" {
...
+}
+
+variable "spacelift_api_key_endpoint" {
+  type        = string
+  description = "Full URL of the Spacelift API endpoint to use, eg. https://demo.app.spacelift.io"
+  default     = "https://example.app.spacelift.io"
```
Comment (jmfontaine) [inline variables.tf:157]: That should not be a default value for the variable but should be passed by the examples to the module.

File: variables.tf [hunk 21]
```diff
@@ -131,3 +131,30 @@ variable "instance_refresh" {
...
+  sensitive   = true
+  description = "Secret corresponding to the Spacelift API key to use"
+}
+
+variable "spacelift_url" {
+  type        = string
+  description = "Full URL of the Spacelift API endpoint to use, eg. https://demo.app.spacelift.io"
+}
+
+variable "local_path" {
```
[the remark this answers is not in the document]
Comment (jmfontaine) [reply variables.tf:160]: `/mnt/workspace/source/examples` works for this repository, but it is unlikely to be present in other repositories. We should use a more common location such as `/tmp` or the local folder (`.`).

**Review by jmfontaine — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 17 above)
[the remark this answers is not in the document]
Comment (jmfontaine) [reply variables.tf]: As discussed yesterday, having the autoscaler repository publish a `latest` package is not as easy as I expected, so let's keep requiring a version number for now.

File: variables.tf (hunk 21 above)
[the remark this answers is not in the document]
Comment (jmfontaine) [reply variables.tf:157]: `/tmp` works in the default runner image and should work for most setups. If it does not, users would be free to override that value.

File: .spacelift/config.yml [hunk 22]
```diff
@@ -3,9 +3,21 @@ module_version: 1.6.0
 tests:
   - name: AMD64-based workerpool
     project_root: examples/amd64
+    environment:
+      TF_VAR_spacelift_api_key_id: "0VOYU49U485BMZZVAWXU59VOW2"
```
Comment (jmfontaine) [inline .spacelift/config.yml:7]:
```suggestion
      TF_VAR_spacelift_api_key_id: "EXAMPLE0VOYU49U485BMZZVAWXU59VOW2"
```

File: .spacelift/config.yml [hunk 23]
```diff
@@ -3,9 +3,21 @@ module_version: 1.6.0
...
+      TF_VAR_spacelift_api_key_secret: "f7anuofh4b6a4e43aplqt49099606de2mzbq4391tj1d3dc9872q23z8fvctu4kh"
```
Comment (jmfontaine) [inline .spacelift/config.yml:8]:
```suggestion
      TF_VAR_spacelift_api_key_secret: "EXAMPLEf7anuofh4b6a4e43aplqt49099606de2mzbq4391tj1d3dc9872q23z8fvctu4kh"
```

File: .spacelift/config.yml [hunk 24]
```diff
@@ -3,9 +3,21 @@ module_version: 1.6.0
...
+      TF_VAR_spacelift_api_key_endpoint: "https://example.app.spacelift.io"
 
   - name: ARM64-based workerpool
     project_root: examples/arm64
+    environment:
+      TF_VAR_spacelift_api_key_id: "0VOYU49U485BMZZVAWXU59VOW2"
```
Comment (jmfontaine) [inline .spacelift/config.yml:14]:
```suggestion
      TF_VAR_spacelift_api_key_id: "EXAMPLE0VOYU49U485BMZZVAWXU59VOW2"
```

File: .spacelift/config.yml [hunk 25]
```diff
@@ -3,9 +3,21 @@ module_version: 1.6.0
...
+      TF_VAR_spacelift_api_key_secret: "f7anuofh4b6a4e43aplqt49099606de2mzbq4391tj1d3dc9872q23z8fvctu4kh"
```
Comment (jmfontaine) [inline .spacelift/config.yml:15]:
```suggestion
      TF_VAR_spacelift_api_key_secret: "EXAMPLEf7anuofh4b6a4e43aplqt49099606de2mzbq4391tj1d3dc9872q23z8fvctu4kh"
```

File: .spacelift/config.yml [hunk 26]
```diff
@@ -3,9 +3,21 @@ module_version: 1.6.0
...
+      TF_VAR_spacelift_api_key_endpoint: "https://example.app.spacelift.io"
 
   - name: Custom IAM Role
     project_root: examples/custom-iam-role
+    environment:
+      TF_VAR_spacelift_api_key_id: "0VOYU49U485BMZZVAWXU59VOW2"
```
Comment (jmfontaine) [inline .spacelift/config.yml:21]:
```suggestion
      TF_VAR_spacelift_api_key_id: "EXAMPLE0VOYU49U485BMZZVAWXU59VOW2"
```

File: .spacelift/config.yml [hunk 27]
```diff
@@ -3,9 +3,21 @@ module_version: 1.6.0
...
+      TF_VAR_spacelift_api_key_secret: "f7anuofh4b6a4e43aplqt49099606de2mzbq4391tj1d3dc9872q23z8fvctu4kh"
```
Comment (jmfontaine) [inline .spacelift/config.yml:22]:
```suggestion
      TF_VAR_spacelift_api_key_secret: "EXAMPLEf7anuofh4b6a4e43aplqt49099606de2mzbq4391tj1d3dc9872q23z8fvctu4kh"
```

File: README.md [hunk 28]
```diff
@@ -2,6 +2,8 @@
 
 Terraform module deploying a Spacelift worker pool on AWS EC2 using an autoscaling group.
 
+This module can optionally deploy a Lambda function to autos-scale the worker pool. The function adds or removes workers depending on the worker pool queue length.
```
Comment (jmfontaine) [inline README.md:5]: I do not see the variable to make the autoscaler optional.

File: README.md [hunk 29]
```diff
@@ -32,6 +34,8 @@ module "my_workerpool" {
 }
 ```
 
+You also need to update [variables](variables.tf) for the Lambda function to set the required `SPACELIFT_API_KEY_ID`, `SPACELIFT_API_KEY_ENDPOINT` and `SPACELIFT_WORKER_POOL_ID` parameters.
```
Comment (jmfontaine) [inline README.md:37]: Users would not update the `variables.tf` file but would provide values for those variables in the different ways Terraform supports.

File: examples/amd64/main.tf [hunk 30]
```diff
@@ -43,4 +43,7 @@ module "this" {
   security_groups = [data.aws_security_group.this.id]
   vpc_subnets     = data.aws_subnets.this.ids
   worker_pool_id  = random_pet.this.id
+  spacelift_api_key_id = var.spacelift_api_key_id
```
Comment (jmfontaine) [inline examples/amd64/main.tf:46]: The arguments should be sorted alphabetically, and the file formatted using the `terraform fmt` command for consistency.

File: examples/arm64/main.tf [hunk 31]
```diff
@@ -66,4 +66,7 @@ module "this" {
   worker_pool_id    = random_pet.this.id
   ami_id            = data.aws_ami.this.id
   ec2_instance_type = "t4g.micro"
+  spacelift_api_key_id = var.spacelift_api_key_id
```
Comment (jmfontaine) [inline examples/arm64/main.tf:69]: The arguments should be sorted alphabetically, and the file formatted using the terraform fmt command for consistency.

File: examples/custom-iam-role/main.tf [hunk 32]
```diff
@@ -65,4 +65,7 @@ module "this" {
   security_groups = [data.aws_security_group.this.id]
   vpc_subnets     = data.aws_subnets.this.ids
   worker_pool_id  = random_pet.this.id
+  spacelift_api_key_id = var.spacelift_api_key_id
```
Comment (jmfontaine) [inline examples/custom-iam-role/main.tf:68]: The arguments should be sorted alphabetically, and the file formatted using the terraform fmt command for consistency.

File: .spacelift/config.yml (hunk 22 above)
Comment (eminalemdar) [reply .spacelift/config.yml]: I somehow completely forgot these after our meeting but added now

File: .spacelift/config.yml (hunk 23 above)
Comment (eminalemdar) [reply .spacelift/config.yml]: I somehow completely forgot these after our meeting but added now

File: .spacelift/config.yml (hunk 24 above)
Comment (eminalemdar) [reply .spacelift/config.yml]: I somehow completely forgot these after our meeting but added now

File: .spacelift/config.yml (hunk 25 above)
Comment (eminalemdar) [reply .spacelift/config.yml]: I somehow completely forgot these after our meeting but added now

File: .spacelift/config.yml (hunk 26 above)
Comment (eminalemdar) [reply .spacelift/config.yml]: I somehow completely forgot these after our meeting but added now

File: .spacelift/config.yml (hunk 27 above)
Comment (eminalemdar) [reply .spacelift/config.yml]: I somehow completely forgot these after our meeting but added now

File: README.md (hunk 28 above)
Comment (eminalemdar) [reply README.md:5]: I've added the parameter

File: README.md (hunk 29 above)
Comment (eminalemdar) [reply README.md]: Updated the instructions and users can add these parameters to the module block

File: examples/amd64/main.tf (hunk 30 above)
Comment (eminalemdar) [reply examples/amd64/main.tf]: Done

File: examples/arm64/main.tf (hunk 31 above)
Comment (eminalemdar) [reply examples/arm64/main.tf]: Done

File: examples/custom-iam-role/main.tf (hunk 32 above)
Comment (eminalemdar) [reply examples/custom-iam-role/main.tf]: Done

File: variables.tf (hunk 21 above)
[the remark this answers is not in the document]
Comment (eminalemdar) [reply variables.tf:163]: Done

File: variables.tf [hunk 33]
```diff
@@ -131,3 +131,43 @@ variable "instance_refresh" {
 locals {
   namespace = "sp5ft-${var.worker_pool_id}"
 }
+
+variable "create_autoscaler_function" {
```
Comment (jmfontaine) [inline variables.tf:135]:
```suggestion
variable "enable_autoscaling" {
```

**Review by jmfontaine — CHANGES_REQUESTED** [review]

**Review by jmfontaine — APPROVED** [review]

**Pull request merged**